### PR TITLE
Allow extending xpath functions with custom JS

### PR DESCRIPTION
**Pull request by garethbowen**
# Description

medic/cht-core#7824

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7824-arbitrary-js/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7824-arbitrary-js/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7824-arbitrary-js/docker-compose/cht-couchdb-clustered.yml)
 
# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by garethbowen** [comment]
Review please @jkuester !

This is the better quality version of the DTree ML prototype that you helped out with, so hopefully the general concept makes sense.

There are two other components to complete this issue - a cht-conf change to make uploading the libs easier, and a cht-docs change to explain the whole thing. I wanted to get another pair of eyes on this to make sure it's roughly right before working on the other changes.

Thanks!

File: api/src/generate-service-worker.js [hunk 1]
```diff
@@ -84,7 +85,7 @@ const writeServiceWorkerFile = async () => {
     dynamicUrlToDependencies: {
       '/': [path.join(webappDirectoryPath, 'index.html')], // Webapp's entry point
       '/medic/login': await getLoginPageContents(),
-      '/medic/_design/medic/_rewrite/': [path.join(webappDirectoryPath, 'appcache-upgrade.html')],
+      '/medic/_design/medic/_rewrite/': [path.join(webappDirectoryPath, 'appcache-upgrade.html')]
```
Comment (garethbowen) [inline api/src/generate-service-worker.js:88]: Note to self: revert this change.

Comment (garethbowen) [inline api/src/generate-service-worker.js:88]: Note to self...

```suggestion
      '/medic/_design/medic/_rewrite/': [path.join(webappDirectoryPath, 'appcache-upgrade.html')],
```

File: api/tests/mocha/generate-service-worker.spec.js [hunk 2]
```diff
@@ -184,7 +194,7 @@ describe('generate service worker', () => {
             '/absolute/path/to/build/static/webapp/fonts/enketo-icons-v2.woff',
             '/absolute/path/to/build/static/webapp/fonts/NotoSans-Bold.ttf',
             '/absolute/path/to/build/static/webapp/fonts/NotoSans-Regular.ttf',
-            '/absolute/path/to/build/static/login/*.{css,js}',
+            '/absolute/path/to/build/static/login/*.{css,js}'
```
Comment (garethbowen) [inline api/tests/mocha/generate-service-worker.spec.js:197]: Note to self.

```suggestion
            '/absolute/path/to/build/static/login/*.{css,js}',
```

File: webapp/src/ts/services/cht-script-api.service.ts [hunk 3]
```diff
@@ -27,16 +30,32 @@ export class CHTScriptApiService {
...
+    await this.getSettings();
+    await this.loadScripts();
   }
 
-  private getSettings() {
-    return this.settingsService
-      .get()
-      .then(settings => this.settings = settings);
+  private async loadScripts() {
+    const extensionLibs = this.settings?.extension_libs;
```
Comment (garethbowen) [inline webapp/src/ts/services/cht-script-api.service.ts:41]: I feel a little uncomfortable about getting this info from `settings`. It means you need to make changes in two places. This will be obscured by cht-conf make it a single command, but it still feels fragile.

The other option would be to replicate the libs into couch, but I feel better about those being part of the service worker like other static resources, rather than having a loader of their own.

If you can think of a better alternative let me know!

File: webapp/tests/karma/ts/services/enketo.service.spec.ts [hunk 4]
```diff
@@ -174,29 +180,31 @@ describe('Enketo service', () => {
     delete window.CHTCore;
   });
 
-  describe('init', () => {
-    it('should init zscore and xpath extensions', async () => {
-      sinon.stub(medicXpathExtensions, 'init');
+  // TODO these tests are breaking because the init is not deterministic -- fix later
```
Comment (garethbowen) [inline webapp/tests/karma/ts/services/enketo.service.spec.ts:183]: Note to self - fix these tests.

File: webapp/tests/karma/ts/services/enketo.service.spec.ts [hunk 5]
```diff
@@ -156,14 +161,13 @@ describe('Enketo service', () => {
...
         },
         { provide: ZScoreService, useValue: zScoreService },
+        { provide: CHTScriptApiService, useValue: chtScriptApiService },
         { provide: TransitionsService, useValue: transitionsService },
         { provide: TranslateService, useValue: translateService },
         { provide: FeedbackService, useValue: feedbackService },
       ],
     });
 
-    service = TestBed.inject(EnketoService);
```
Comment (garethbowen) [inline webapp/tests/karma/ts/services/enketo.service.spec.ts:165]: This was being constructed too soon, and the enketo service constructor was calling `init()` which meant services were being called before the tests had a chance to stub them. By moving this down into the test we can stub them before they get initialised.

File: webapp/src/ts/services/enketo.service.ts [hunk 6]
```diff
@@ -76,10 +78,15 @@ export class EnketoService {
   }
 
   private init() {
+    if (this.inited) {
+      return this.inited;
+    }
```
Comment (garethbowen) [inline webapp/src/ts/services/enketo.service.ts:83]: This avoids a race condition where you could get double initialisation.

File: api/src/services/extension-libs.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    return Object.entries(doc._attachments).map(([ name, attachment ]) => ({ name, attachment }));
+  },
+  get: async (libName) => {
+    const doc = await getLibsDoc();
+    const attachment = doc && doc._attachments && doc._attachments[libName];
+    if (attachment) {
+      return {
+        data: attachment.data,
+        contentType: attachment.content_type
+      };
```
Comment (jkuester) [inline api/src/services/extension-libs.js:33]: At the risk of over-architecting, I wonder if we should consider building in the concept of an `extension_type` right from the beginning (the value for these would be something like `xpath_function`)?  My thought is that in the future we will probably want to add other types of extensions besides just custom xpath functions. It might make sense to include those extensions as attachments to this `extension-libs` doc too, but it would be good to be able to differentiate.

**Review by jkuester — CHANGES_REQUESTED** [review]
Super excited about this funcitonality!  Some thoughs inline below.

File: api/src/routing.js [hunk 8]
```diff
@@ -250,6 +252,7 @@ app.all('/+admin(/*)?', authorization.handleAuthErrors, authorization.offlineUse
 
 app.use(express.static(environment.staticPath));
 app.use(express.static(environment.webappPath));
+app.get('/extension-libs/:name', extensionLibs.get);
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/routing.js:255]: Yeah I'm not 100% happy with this name, but naming is hard. I feel `extensions` is too generic, and as we get into more extensibility may complicate things. This is exclusively for separate JS files, so libs felt write, but maybe `extension-scripts`? Maybe it's the `extension` bit that's the problem and it should be `configured-scripts`? I haven't found anything I really like yet...

File: api/src/controllers/extension-libs.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+const serverUtils = require('../server-utils');
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/controllers/extension-libs.js:1]: Good call. Done.

File: api/src/controllers/extension-libs.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+const extensionLibs = require('../services/extension-libs');
+
+module.exports = {
+  get: async (req, res) => {
+    const lib = await extensionLibs.get(req.params.name);
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/controllers/extension-libs.js:6]: Definitely, thanks. Done.

File: api/src/services/extension-libs.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
+const db = require('../db');
+
+const DOC_ID = 'extension-libs';
+
+const getLibsDoc = async () => {
+  try {
+    return await db.medic.get(DOC_ID, { attachments: true });
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/services/extension-libs.js:7]: Good point. I'm naturally inclined to try and make it explicit, but you're right, it's not necessary at all.

File: api/src/services/extension-libs.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+module.exports = {
+  isLibChange: (change) => (change && change.id) === DOC_ID,
+  getAll: async () => {
+    const doc = await getLibsDoc();
+    if (!doc || !doc._attachments) {
+      return [];
+    }
+    return Object.entries(doc._attachments).map(([ name, attachment ]) => ({ name, attachment }));
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/services/extension-libs.js:24]: Absolutely. I sort of missed that I was exposing the internal structure there. I've reformatted so they both use the same format and hidden the CouchDB format completely.

File: api/tests/mocha/services/extension-libs.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
+const sinon = require('sinon');
+const chai = require('chai');
+
+const service = require('../../../src/services/extension-libs');
+const db = require('../../../src/db');
+
+describe.only('Extension Libs service', () => {
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/tests/mocha/services/extension-libs.js:7]: Hehe. Yeah I noticed that this morning when I tried to run all unit tests... Thanks!

File: api/tests/mocha/services/extension-libs.js [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+    it('throws anything else', async () => {
+      dbGet.throws({ status: 403 });
+      try {
+        await service.getAll();
+      } catch(e) {
+        chai.expect(e.status).to.equal(403);
+        return;
+      }
+      throw new Error('Expected error to be thrown');
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/tests/mocha/services/extension-libs.js:51]: I think I like my way better. They both correctly fail the test which is the important thing, but when you use `chai.expect.fail` it throws an Error under the hood, which gets caught by the `catch(e)`. Then it fails with a hard to diagnose message because it's actually failing on the `chai.expect` line... eg:

Using `chai.expect.fail`:

```
  1) Extension Libs service
       get
         throws anything else:
     AssertionError: expected undefined to equal 403
      at Context.<anonymous> (api/tests/mocha/services/extension-libs.spec.js:109:34)
```

Using `throw new Error` outside of the `catch`:

```
  1) Extension Libs service
       get
         throws anything else:
     Error: Expected error to be thrown
      at Context.<anonymous> (api/tests/mocha/services/extension-libs.spec.js:112:13)
```

File: api/tests/mocha/services/extension-libs.js [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+      const actual = service.isLibChange({ id: 'extension-libs' });
+      chai.expect(actual).to.be.true;
+    });
+
+  });
+
+  describe('getAll', () => {
+
+    it('handles 404', async () => {
+      dbGet.throws({ status: 404 });
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/tests/mocha/services/extension-libs.js:38]: Done.

File: api/src/services/config-watcher.js [hunk 16]
```diff
@@ -163,6 +168,10 @@ const listen = () => {
       if (change.id === 'branding') {
         return handleBrandingChanges();
       }
+
+      if (extensionLibs.isLibChange(change)) {
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/services/config-watcher.js:172]: Done.

File: webapp/src/js/enketo/medic-xpath-extensions.js [hunk 17]
```diff
@@ -176,6 +178,12 @@ module.exports = {
       const months = d2Moment.diff(d1Moment, 'months');
       return XPR.number(months);
     },
+    'extension-lib': function() {
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/enketo/medic-xpath-extensions.js:181]: Ooh, I like the CHT namespace. This also applies to all the other functions in xpath-extensions, right? Though, obviously any change needs to be done in a backwards compatible way.

I'll make the change for to `cht:extension-lib` (name to be confirmed from discussion above) and we can look at migrating the rest soon.

File: webapp/src/js/enketo/medic-xpath-extensions.js [hunk 18]
```diff
@@ -176,6 +178,12 @@ module.exports = {
...
+      const args = Array.from(arguments);
+      const libId = args.shift();
+      const lib = chtScriptApi.v1.getExtensionLib(libId.v);
+      return lib.apply(null, args);
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/enketo/medic-xpath-extensions.js:185]: I agree it's a shame, but I didn't want to start down a rabbit hole of having to handle the different types. Anything we hardcode here restricts what can be done in a custom way.

File: webapp/src/js/enketo/medic-xpath-extensions.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/enketo/medic-xpath-extensions.js:181]: Done.

File: webapp/src/ts/services/cht-script-api.service.ts [hunk 19]
```diff
@@ -27,16 +30,32 @@ export class CHTScriptApiService {
...
+    if (extensionLibs && extensionLibs.length) {
+      return Promise.all(extensionLibs.map(name => this.loadScript(name)));
+    }
+  }
+
+  private async loadScript(name) {
+    try {
+      const src = await this.http.get('/extension-libs/' + name, { responseType: 'text' }).toPromise();
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/ts/services/cht-script-api.service.ts:49]: Replaced. It's just so ugly...

**Comment by garethbowen** [comment]
> Hmm, yes I agree that it seems less than ideal to have to set the extension names in the settings. (and cht-conf doing it automatically does make it less work for the developer trying to configure the extension, but makes it even more crazy from a coding perspective since now we require the double-config AND have to maintain code to automate setting the double-config.). Honestly, is there a reason we cannot just have an api endpoint that returns all the extensions, or even just metadata about all the extensions (like their name...)?

Couldn't reply in line so replying here...

The problem is that making a request like that doesn't work offline. The result of that call needs to be stored somewhere. We could store it somewhere else but that doesn't really solve the problem - we're still storing it in two places.

File: webapp/src/ts/services/enketo.service.ts [hunk 20]
```diff
@@ -76,10 +78,15 @@ export class EnketoService {
...
     return this.zScoreService
       .getScoreUtil()
       .then((zscoreUtil) => {
-        medicXpathExtensions.init(zscoreUtil, toBik_text, moment);
+        return this.chtScriptApiService.getApi().then(api => {
+          medicXpathExtensions.init(zscoreUtil, toBik_text, moment, api);
+        });
       })
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/ts/services/enketo.service.ts:90]: Done.

File: api/src/services/extension-libs.js (hunk 7 above)
Comment (garethbowen) [reply api/src/services/extension-libs.js:33]: Trying to be forwards compatible is hard... I don't think we need to change anything in code, because we can change that easily enough when we want to use the `extension_type` property. To be safe we _could_ store something in the doc, so when that change happens we can just add a new type and there's no migration. However because this is only modifying one doc it's easy enough to migrate, or just handle the "unknown" case by defaulting to `xpath_function`.

I'm reluctant to change anything at this stage because I don't know enough about what other extensions will look like...

File: api/tests/mocha/services/extension-libs.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (jkuester) [reply api/tests/mocha/services/extension-libs.js]: Ohhh yeah that is better failure behavior!  The whole setup, though, still feels primitive....  I was reading a little bit and with `chai-as-promised`, we can do this:

```js
      await chai.expect(service.getAll()).to.eventually.be.rejected
        .with.property('status', 403);
```

That seems to give good failure errors both in the case where the expected status is not found and when the promise is not rejected.  

I am fine leaving things how you have it, but I am going to try and use `chai-as-promised` more in the future....

File: api/src/routing.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jkuester) [reply api/src/routing.js:255]: > This is exclusively for separate JS files, so libs felt write

Ahh, I am understanding more of your reasoning behind that name!  This is pure opinion, but to me `extension-scripts > extension-libs > configured-scripts` but honestly it's a fairly weak preference and I am not sure it would be worth going back and re-naming everything.  I am fine with `extension-libs` if we cannot come up with anything else that is clearly better!

**Comment by jkuester** [comment]
> The problem is that making a request like that doesn't work offline. The result of that call needs to be stored somewhere.

I am sure this is my ignorance of web development here, but why can't we include the data for that call in the service worker just like we did for `/extension-libs`?

File: api/src/services/extension-libs.js (hunk 7 above)
Comment (jkuester) [reply api/src/services/extension-libs.js]: Fair.  Probably best to not try to blindly solve problems we don't have yet....

**Comment by garethbowen** [comment]
> I am sure this is my ignorance of web development here, but why can't we include the data for that call in the service worker just like we did for /extension-libs?

Hmm... Almost like a `/extension-libs/index.js`. I'll have a think, but it _should_ work just fine. Good thinking.

**Comment by garethbowen** [comment]
@jkuester I think I've made all those updates, including removing the settings doc requirement. I also added an additional change to ensure nobody except an admin can update the `extension-libs` doc. The tests are failing because of the leap year problem that is fixed in `master` - one you approve this PR I'll rebase and send off to AT.

File: webapp/src/ts/services/cht-script-api.service.ts [hunk 21]
```diff
@@ -27,16 +32,33 @@ export class CHTScriptApiService {
...
+  private async init() {
     this.watchChanges();
     this.userCtx = this.sessionService.userCtx();
-    return this.getSettings();
+    await Promise.all([ this.getSettings(), this.loadScripts() ]);
+  }
+
+  private async loadScripts() {
+    const request = this.http.get<String[]>('/extension-libs', { responseType: 'json' });
+    const extensionLibs = await lastValueFrom(request);
```
Comment (jkuester) [inline webapp/src/ts/services/cht-script-api.service.ts:43]: Should we wrap this in a try-catch?  (Looks like we are doing that in the other places we call `http.get`....)

**Review by jkuester — APPROVED** [review]
This is great!  Just a few minor questions about some error handling.

File: webapp/src/ts/services/cht-script-api.service.ts (hunk 21 above)
Comment (garethbowen) [reply webapp/src/ts/services/cht-script-api.service.ts:43]: Done!

File: webapp/src/js/enketo/medic-xpath-extensions.js [hunk 22]
```diff
@@ -176,6 +178,12 @@ module.exports = {
       const months = d2Moment.diff(d1Moment, 'months');
       return XPR.number(months);
     },
+    'cht:extension-lib': function() {
+      const args = Array.from(arguments);
+      const libId = args.shift();
+      const lib = chtScriptApi.v1.getExtensionLib(libId.v);
+      return lib.apply(null, args);
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/enketo/medic-xpath-extensions.js:185]: Absolutely! Good spotting. Caught and thrown and added a unit test.
